In [42]:
import numpy as np
import pandas as pd
import os
import re
#https://www.nltk.org/
from sklearn.feature_extraction.text import CountVectorizer
import tensorflow as tf
import ast
from matplotlib import pyplot as plt

In [43]:
# from keybert import KeyBERT
# kw_model = KeyBERT()

In [44]:
#Extract the dataset we are interested in, synopsis, name, and rating.
df = pd.read_csv('DataSet/animes.csv')
df = df[['title','synopsis','score']]
df.dropna(inplace=True)
df = df[~df.title.duplicated(keep='first')]
df.head()

title  \
0           Haikyuu!! Second Season   
1           Shigatsu wa Kimi no Uso   
2                     Made in Abyss   
3  Fullmetal Alchemist: Brotherhood   
4  Kizumonogatari III: Reiketsu-hen   

                                            synopsis  score  
0  Following their participation at the Inter-Hig...   8.82  
1  Music accompanies the path of the human metron...   8.83  
2  The Abyss—a gaping chasm stretching down into ...   8.83  
3  "In order for something to be obtained, someth...   9.23  
4  After helping revive the legendary vampire Kis...   8.83

In [45]:
#Method one for generating X vectors: hash.
from sklearn.feature_extraction.text import HashingVectorizer
X = df['synopsis'].to_numpy()
y = df['score'].to_numpy()
y = np.rint(y)
print(len(y))
vectorizer = HashingVectorizer(n_features=100, norm = None, alternate_sign = False)
hashed_X = vectorizer.fit_transform(X)
#print(hashed_X)
hashed_X.shape

15192


(15192, 100)

In [46]:
#Building training sets and test sets.
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(hashed_X, y, test_size = 0.2)

In [49]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
model = DecisionTreeClassifier(max_depth=100)
model.fit(X_train, y_train)
pred = model.predict(X_test)

score = accuracy_score(y_test, pred)
print("Score: " + str(score))

Score: 0.3504442250740375


Accuracy of about 35%, which is better than guessing randomly (10%), but not very good for practical predictions.

Pros of this method: Easy to implement

Cons: Not interpretable, I have no idea what type of synopsis correspond to the predictions made.